### PR TITLE
Rewrite autopropose in Python as a step towards complicated scenarios

### PR DESCRIPTION
**Pull request by adaszko**
I think this should be run in a single network for now.  `sandboxnet` perhaps?  Any easy way to deploy it just there?

https://rchain.atlassian.net/browse/RCHAIN-3807

### CODE REVIEW

File: autopropose.service [hunk 1]
```diff
@@ -4,7 +4,7 @@ Requires=docker.service
 After=docker.service
 
 [Service]
-ExecStart=/opt/rchain-testnet-node/autopropose
+ExecStart=/opt/rchain-testnet-node/autoproposeng.py
```
Comment (woky) [inline autopropose.service]: Keep the filename please.

Comment (adaszko) [reply autopropose.service]: > Keep the filename please.
OK.

> This is effectively 1:1 rewrite from Bash to Python.
I need it to evolve autopropose into what's been proposed at https://rchain.atlassian.net/wiki/spaces/OP/pages/752058391/Network+Hardening

> What complicated scenarios do you have in mind and how exactly does this help?
The "implementation" section at https://rchain.atlassian.net/wiki/spaces/OP/pages/752058391/Network+Hardening

**Comment by woky** [comment]
> > What complicated scenarios do you have in mind and how exactly does this help?
> 
> The "implementation" section at https://rchain.atlassian.net/wiki/spaces/OP/pages/752058391/Network+Hardening

OK, I get it, that you want to rewrite it in Python to implement more complex scenario. But how does **1:1 rewrite from Bash to Python** actually gets you closer to that? The paragraph you linked doesn't explain that. This is is half assed:

```
def parse_bash_conf_file(file_name):
    command = ['env', '-i', 'bash', '-c', 'source {} && env'.format(file_name)]
    proc = subprocess.Popen(command, stdout=subprocess.PIPE)
    result = {}
    for line in proc.stdout:
        (key, _, value) = line.partition("=")
        result[key] = value
    proc.communicate()
    return result
```
```
         rnode_container_hash = subprocess.check_output(['docker', 'ps', '--quiet', '--filter=name=^rnode$']) 
```

You can do better with Python.

**Comment by adaszko** [comment]
> OK, I get it, that you want to rewrite it in Python to implement more complex scenario. But how does 1:1 rewrite from Bash to Python actually gets you closer to that?

Step 1) Rewrite autopropose in Python.
Step 2) Incorporate deploy script into it.
Step 3) Incorporate propose script into it.
Step 4) Implement vault transfers measuring time until finalization.  Report that time to InfluxDB.
Step 5) Implement measuring the above with latency injection with `tsc` command.
Step 6) As above, but with cutting off nodes.

> This is is half assed:

Mind your language.  It's unprofessional.

**Comment by woky** [comment]
> Step 1) Rewrite autopropose in Python.

You did *NOT* _rewrite_ it into Python. You merely translated the Bash script into Python. I'm not going to approve commits that merely translate files from language A to B. Submit meaningful commit that clearly shows an improvement or implement what you want in what's available. This is garbage.

**Comment by adaszko** [comment]
> Let's merge this.

I appreciate the decision but there are some unresolved questions.  I don't want to it be deployed to all of: devnet, devnet2, testnet, sandboxnet and I believe that's what will happen once it's merged.  I only need to deploy to one of the networks (sandboxnet?) so that I can experiment there and not disrupt other people's work in the remaining networks.  Things like long propose times investigation.

**Comment by woky** [comment]
> I appreciate the decision but there are some unresolved questions. I don't want to it be deployed to all of: devnet, devnet2, testnet, sandboxnet and I believe that's what will happen once it's merged.

Testnet uses master branch. It'll be deployed to the rest though. See `bootstrap*` files. `REPO_BRANCH` defaults to `master` if not set.

> I only need to deploy to one of the networks (sandboxnet?) so that I can experiment there and not disrupt other people's work in the remaining networks.

You can modify Git remote on a machine. Say you want to use repository https://github.com/foobar/rchain-testnet-node branch `testbranch`. Set `REPO_URL` and `REPO_BRANCH` accordingly in `bootstrap.sandboxnet` in that branch (this is just temporary but has to be committed and pushed), and then do on a node:
```
cd /opt/rchain-testnet-node
git remote add foobar https://github.com/foobar/rchain-testnet-node
git fetch foobar
git checkout foobar/testbranch
./update-run-setup
# use the following if branch switch fails
# env FORCE_UPDATE_1 ./update-run-setup
```
To do it on multiple nodes I use https://github.com/lilydjwg/pssh but you can use any SSH multiplexer (e.g. GNU parallel, Ansible, shmux, pdsh).

**Comment by woky** [comment]
LGTM. Do you mind just replacing previous `autopropose`, that is to remove `ng` and `.py` suffixes?

**Comment by adaszko** [comment]
> LGTM. Do you mind just replacing previous autopropose, that is to remove ng and .py suffixes?

The only instance of autoproposeng is `REPO_BRANCH`.

As I said, autopropose.py runs only in sandbox for now, so I have to leave autopropose script in other places.  Once it is sufficiently robust and I'm sure it behaves well, I will roll it out everywhere.  So for now, the autopropose script needs to stay in place.

**Comment by woky** [comment]
FWIW, I don't think you need to _support_ `deploy` and `propose` script. They were meant to be used just for poking the network from Rundeck. They were also used in previous `autopropose` just to not avoid duplication on code. You can use gRPC directly (e.g. `propose`/`deploy` methods in pyrchain) if you feel like it.

**Comment by adaszko** [comment]
> FWIW, I don't think you need to support deploy and propose script. They were meant to be used just for poking the network from Rundeck. They were also used in previous autopropose just to not avoid duplication on code. You can use gRPC directly (e.g. propose/deploy methods in pyrchain) if you feel like it.

OK, good to know.  I did intend to use direct gRPC calls in the end.

**Comment by woky** [comment]
Also, I think that this `autopropose` thing might survive MainNet launch. I don't know of any plans to make RNode propose automatically, so what you're doing could be running on MainNet (with deploy disabled of course).

**Comment by adaszko** [comment]
> Also, I think that this `autopropose` thing might survive MainNet launch. I don't know of any plans to make RNode propose automatically, so what you're doing could be running on MainNet (with deploys disabled of course). So feel free to make bigger changes to align with that. I guess finalization will need to be handled as well. We need to ask @KentShikama and others.

Finalization is definitely a big part of this.  Daniyar is helping me with it.

Is this PR good to go then?

**Comment by woky** [comment]
> Is this PR good to go then?
LGTM.

**Pull request merged**